support/k8s/k8s-workload-registrar/main.go

```diff
@@ -1,73 +1,210 @@
-
-	controller := NewController(ControllerConfig{
-		Log:           log,
-		R:             registration.NewRegistrationClient(serverConn),
-		TrustDomain:   config.TrustDomain,
-		Cluster:       config.Cluster,
-		PodLabel:      config.PodLabel,
-		PodAnnotation: config.PodAnnotation,
-	})
 
-	log.Info("Initializing registrar")
-	if err := controller.Initialize(ctx); err != nil {
-		return err
+	mode := controllers.PodReconcilerModeServiceAccount
+	value := ""
+	if len(config.PodLabel) > 0 {
+		mode = controllers.PodReconcilerModeLabel
+		value = config.PodLabel
+	}
+	if len(config.PodAnnotation) > 0 {
+		mode = controllers.PodReconcilerModeAnnotation
+		value = config.PodAnnotation
+	}
+	if err = controllers.NewPodReconciler(
+		mgr.GetClient(),
+		ctrl.Log.WithName("controllers").WithName("Pod"),
+		mgr.GetScheme(),
+		config.TrustDomain,
+		rootID,
+		spireClient,
+		mode,
+		value,
+		config.ClusterDNSZone,
+		config.AddPodDNSNames,
+	).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "Unable to create controller", "controller", "Pod")
+		os.Exit(1)
 	}
 
-	server, err := NewServer(ServerConfig{
-		Log:                            log,
-		Addr:                           config.Addr,
-		Handler:                        NewWebhookHandler(controller),
-		CertPath:                       config.CertPath,
-		KeyPath:                        config.KeyPath,
-		CaCertPath:                     config.CaCertPath,
-		InsecureSkipClientVerification: config.InsecureSkipClientVerification,
-	})
-	if err != nil {
-		return err
+	// +kubebuilder:scaffold:builder
+
+	setupLog.Info("Starting manager")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		setupLog.Error(err, "Problem running manager")
+		os.Exit(1)
 	}
-	return server.Run(ctx)
+}
+
+type SpiffeLogWrapper struct {
+	delegate logr.Logger
+}
+
+func (slw SpiffeLogWrapper) Debugf(format string, args ...interface{}) {
+	slw.delegate.V(1).Info(fmt.Sprintf(format, args...))
+}
+func (slw SpiffeLogWrapper) Infof(format string, args ...interface{}) {
+	slw.delegate.Info(fmt.Sprintf(format, args...))
+}
+func (slw SpiffeLogWrapper) Warnf(format string, args ...interface{}) {
+	slw.delegate.Info(fmt.Sprintf(format, args...))
+}
+func (slw SpiffeLogWrapper) Errorf(format string, args ...interface{}) {
+	slw.delegate.Info(fmt.Sprintf(format, args...))
+}
+
+func ConnectSpire(ctx context.Context, log logr.Logger, serverAddress, agentSocketPath string) (registration.RegistrationClient, error) {
+	var conn *grpc.ClientConn
+	var err error
+
+	if agentSocketPath == "" {
+		log.Info("Connecting to workload API without security", "serverAddress", serverAddress)
+		conn, err = grpc.DialContext(ctx, serverAddress, grpc.WithInsecure())
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		source, err := workloadapi.NewX509Source(ctx, workloadapi.WithClientOptions(workloadapi.WithAddr("unix://"+agentSocketPath)), workloadapi.WithClientOptions(workloadapi.WithLogger(SpiffeLogWrapper{log})))
```

Comment:
This is main, so it probably doesn't matter practically, but technically this source should be closed whenever updates are no longer required from the Workload API (i.e. when the grpc conn is closed)